### ISSUE TITLE
Update internal `torch.norm` calls to `torch.linalg.norm`

### ISSUE THREAD

**Issue by kurtamohler** [pytorch/pytorch#49907]
Since `torch.norm` is deprecated in favor of `torch.linalg.norm`, all internal calls should be updated.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @heitorschueroff

**Comment by ngimel** [comment]
Last time I checked, .norm calls (with p!=2) in clip_grad_norm could not be updated to use torch.linalg.norm because it does not provide the necessary functionality?

**Comment by kurtamohler** [comment]
@ngimel, I think the functionality you're talking about is the ability to perform a vector norm on an n-dimensional tensor. `torch.norm` supports this, but `torch.linalg.norm` does not, since the norm type (vector vs. matrix) is determined by the number of input dimensions. The second `torch.norm` call on this line has an n-dimensional input: https://github.com/pytorch/pytorch/blob/master/torch/nn/utils/clip_grad.py#L36

However, we actually can use `torch.linalg.norm` as long as we flatten the input first. Replacing the line I linked above with the following line seems to work fine, and `test_clip_grad_norm` from `test_nn.py` passes. The test fails if we don't perform the `flatten` operation.

```
total_norm = torch.linalg.norm(torch.stack([torch.linalg.norm(p.grad.detach().flatten(0, -1), norm_type).to(device) for p in parameters]), norm_type)
```

**Comment by ngimel** [comment]
Sorry, `.flatten` call incurs a significant overhead, so (and especially since currently it's not needed) please don't do this.

**Comment by mruberry** [comment]
We should fix this design issue so we can replace all calls with torch.linalg.norm (and use this issue to discuss how to do that).

One option would be to add a kwarg like "norm_type" or "vector_norm" or "matrix_norm" that could specify whether we're taking a vector or matrix norm.

**Comment by kurtamohler** [comment]
I see, I didn't realize flatten was so expensive. I like the idea of adding an argument to allow n-dimensional inputs. It might only make sense to do this for the vector norms. For vector norm, it's simple to interpret the n-D input as a 1-D vector. But for matrix norm, we would probably have to let the user tell us which set of dimensions to interpret as dim 0 and which set to interpret as dim 1.

Only supporting n-D inputs for vector norms should cover all our bases as well. `torch.norm` has two matrix norms, Frobenius and Nuclear. Nuclear does not support greater than 2-D inputs. Frobenius does support more than two dimensions, but it's the same thing as a flattened ord=2 vector norm.

We could add a bool arg called `flattened_vector_norm`, and document its behavior as something like "If true, the input tensor is interpreted as a flattened vector, and the overhead of a flatten operation is avoided"

**Comment by mruberry** [comment]
Adding a bool arg for flattening sounds good. We can bikeshed on the name later but maybe just "flatten" is the thing? Since really it's just fusing the flatten operation. 

I think we can move forward with that approach and validate the performance.

**Comment by mruberry** [comment]
Updated title of this issue to reflect our latest thinking.

**Comment by lezcano** [comment]
Norm is now implemented in terms of `linalg.vector_norm` and `linalg.matrix_norm`.
